dotnet/src/SemanticKernel/Planning/ConditionalFlowHelper.cs

```diff
@@ -0,0 +1,341 @@
+. If condition has an error, valid will be false and property ""reason"" will have all detail why.
+. If condition is valid, update reason with the current evaluation detail.
+. Return ""condition"" as true or false depending on the condition evaluation
+
+Variables Example:
+x = 2
+y = ""adfsdasfgsasdddsf""
+z = 100
+w = ""sdf""
+
+Given Example:
+$x equals 1 and $y contains 'asd' or not ($z greaterthan 10)
+
+Reason Example:
+(x == 1 ∧ (y contains ""asd"") ∨ ¬(z > 10))
+(TRUE ∧ TRUE ∨ ¬ (FALSE))
+(TRUE ∧ TRUE ∨ TRUE)
+TRUE
+
+Variables Example:
+24hour = 11
+
+Given Example:
+$24hour equals 1 and $24hour greaterthan 10
+
+Response Example:
+{ ""valid"": true, ""condition"": false, ""reason"": ""(24hour == 1 ∧ 24hour > 10) = (FALSE ∧ TRUE) = FALSE"" }
+
+Variables Example:
+24hour = 11
+
+Given Example:
+Some condition
+
+Response Example:
+{ ""valid"": false, ""reason"": ""<detail why>"" }
+
+Variables Example:
+a = 1
+b = undefined
+c = ""dome""
+
+Given Example:
+(a is not undefined) and a greaterthan 100 and a greaterthan 10 or (a equals 1 and a equals 10) or (b is undefined and c is not undefined)
+
+Response Example:
+{ ""valid"": true, ""condition"": true, ""reason"": ""((a is not undefined) ∧ a > 100 ∧ a > 10) ∨ (a == 1 ∧ a == 10) ∨ (b is undefined ∧ c is not undefined) = ((TRUE ∧ FALSE ∧ FALSE) ∨ (TRUE ∧ FALSE) ∨ (TRUE ∧ TRUE)) = (FALSE ∨ FALSE ∨ TRUE) = TRUE"" }
+
+Variables:
+{{$ConditionalVariables}}
+
+Given:
+{{$IfCondition}}
+
+Response: ";
+
+    #endregion
+
+    internal const string ReasonIdentifier = "Reason:";
+    internal const string NoReasonMessage = "No reason was provided";
+
+    private readonly ISKFunction _ifStructureCheckFunction;
+    private readonly ISKFunction _evaluateConditionFunction;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ConditionalFlowHelper"/> class.
+    /// </summary>
+    /// <param name="kernel"> The kernel to use </param>
+    /// <param name="completionBackend"> A optional completion backend to run the internal semantic functions </param>
+    internal ConditionalFlowHelper(IKernel kernel, ITextCompletion? completionBackend = null)
+    {
+        this._kernel = kernel;
+        this._ifStructureCheckFunction = kernel.CreateSemanticFunction(
+            IfStructureCheckPrompt,
+            skillName: "PlannerSkill_Excluded",
+            description: "Evaluate if an If structure is valid and returns TRUE or FALSE",
+            maxTokens: 100,
+            temperature: 0,
+            topP: 0.5);
+
+        this._evaluateConditionFunction = kernel.CreateSemanticFunction(
+            EvaluateConditionPrompt,
+            skillName: "PlannerSkill_Excluded",
+            description: "Evaluate a condition group and returns TRUE or FALSE",
+            maxTokens: 100,
+            temperature: 0,
+            topP: 0.5);
+
+        if (completionBackend is not null)
+        {
+            this._ifStructureCheckFunction.SetAIService(() => completionBackend);
```

Comment:
Is this required? Doesn't tihs come from the `kernel` instance used to create the semantic function?

Reply:
@lemillermicrosoft this is required for Unit Testing, so I inject the Mock of the service to mock the LLM response

Reply:
Ahh hmm okay. I get a little hesitant having code that's *only* for unit testing. It's not a huge deal though. I was working on unit tests for the WIP I have and while it's a little high on startup cost, I thought it'd be possible to mock out the kernel object enough to enable it, e.g. mock out `RunAsync` or `InvokeAsync`

Reply:
@lemillermicrosoft I thought at first this approach but doing so I realized would not be able to unittest the SemanticFunctions logic with RunAsync.

Same for the ISKFunction.Invoke because those are instantiated inside the Concrete Flow Helper and to be able to do that I would need to move those to be IOC injected thru the ConditionalFlowHelper contructor...

Reply:
For Unit Tests, I think mocking `RunAsync` might be enough. Fine to revisit later though.